pkg/eestream/encode.go

```diff
@@ -38,71 +38,103 @@ type ErasureScheme interface {
 }
 
 type encodedReader struct {
-	r               io.Reader
-	es              ErasureScheme
-	cv              *sync.Cond
-	inbuf           []byte
-	outbufs         [][]byte
-	piecesRemaining int
-	err             error
+	r       io.Reader
+	es      ErasureScheme
+	inbuf   []byte
+	outbufs [][]byte
+	chans   map[int]chan []byte
+	err     error
 }
 
 // EncodeReader will take a Reader and an ErasureScheme and return a slice of
 // Readers
-func EncodeReader(r io.Reader, es ErasureScheme) []io.Reader {
+func EncodeReader(r io.Reader, es ErasureScheme, maxBufferMemory int) []io.Reader {
 	er := &encodedReader{
 		r:       r,
 		es:      es,
-		cv:      sync.NewCond(&sync.Mutex{}),
 		inbuf:   make([]byte, es.DecodedBlockSize()),
 		outbufs: make([][]byte, es.TotalCount()),
+		chans:   make(map[int]chan []byte, es.TotalCount()),
 	}
 	readers := make([]io.Reader, 0, es.TotalCount())
 	for i := 0; i < es.TotalCount(); i++ {
-		er.outbufs[i] = make([]byte, 0, es.EncodedBlockSize())
 		readers = append(readers, &encodedPiece{
 			er: er,
 			i:  i,
 		})
 	}
+	if maxBufferMemory < 0 {
+		er.err = Error.New("negative max buffer memory")
+		return readers
+	}
+	chanSize := maxBufferMemory / (es.TotalCount() * es.EncodedBlockSize())
+	if chanSize < 1 {
+		chanSize = 1
+	}
+	for i := 0; i < es.TotalCount(); i++ {
+		er.chans[i] = make(chan []byte, chanSize)
+	}
+	go er.fillBuffer()
 	return readers
 }
 
-func (er *encodedReader) wait() (err error) {
-	// have we already failed? just return that
-	if er.err != nil {
-		return er.err
+func (er *encodedReader) fillBuffer() {
+	defer er.closeChannels()
+	for {
+		_, err := io.ReadFull(er.r, er.inbuf)
+		if err != nil {
+			return
```

Comment:
can we log this error? this will be surprising to someone down the road who is trying to figure out what's going on. cool we close the channels and return, but i'd love for all non-nil errors to somehow wind up getting logged somewhere either via returning up the stack or logging directly

Reply:
My plan is to send an error message to the channel just before closing it. This way the error will be delivered to the respective io.Reader. Is this good enough?